File: tests/test_analytics.py
"""本地统计分析测试"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from analytics import (
    compute_basic_stats,
    top_authors,
    top_categories,
    time_of_day_distribution,
    top_authors_by_completion,
    daily_trend,
)


def _make_video(title, author, category, view_at, duration, progress):
    return {
        "标题": title,
        "BV号": "BV1test",
        "UP主": author,
        "分类": category,
        "观看时间": view_at,
        "总时长(秒)": duration,
        "已观看(秒)": progress,
        "类型": "video",
    }


def _make_live(author, view_at):
    return {
        "标题": "直播标题",
        "房间号": "123",
        "主播": author,
        "分类": "未开播",
        "观看时间": view_at,
        "类型": "live",
    }


def _make_article(author, category, view_at):
    return {
        "标题": "专栏标题",
        "文章ID": "456",
        "作者": author,
        "分类": category,
        "观看时间": view_at,
        "类型": "article",
    }


class TestBasicStats:
    def test_empty(self):
        stats = compute_basic_stats([])
        assert stats["total_records"] == 0
        assert stats["total_watch_time_text"] == "0分钟"

    def test_counts_and_watch_time(self):
        records = [
            _make_video("v1", "UP科技", "科技", "2024-06-20 10:00:00", 600, 300),
            _make_video("v2", "UP科技", "科技", "2024-06-20 11:00:00", 600, 600),
            _make_live("主播A", "2024-06-20 12:00:00"),
            _make_article("作者A", "数码", "2024-06-20 13:00:00"),
        ]
        stats = compute_basic_stats(records)
        assert stats["total_records"] == 4
        assert stats["total_videos"] == 2
        assert stats["total_lives"] == 1
        assert stats["total_articles"] == 1
        assert stats["unique_authors"] == 3
        assert stats["avg_completion"] == "75.0%"


class TestTopAuthors:
    def test_order_by_count(self):
        records = [
            _make_video("v1", "A", "科技", "2024-06-20 10:00:00", 600, 600),
            _make_video("v2", "A", "科技", "2024-06-20 11:00:00", 600, 600),
            _make_video("v3", "B", "科技", "2024-06-20 12:00:00", 600, 600),
        ]
        authors = top_authors(records)
        assert authors[0]["name"] == "A"
        assert authors[0]["count"] == 2
        assert authors[1]["name"] == "B"


class TestTopCategories:
    def test_group_by_category(self):
        records = [
            _make_video("v1", "A", "音乐", "2024-06-20 10:00:00", 600, 600),
            _make_video("v2", "B", "音乐", "2024-06-20 11:00:00", 600, 600),
            _make_video("v3", "C", "科技", "2024-06-20 12:00:00", 600, 600),
        ]
        cats = top_categories(records)
        assert cats[0]["name"] == "音乐"
        assert cats[0]["count"] == 2
        assert cats[1]["name"] == "科技"


class TestTimeOfDay:
    def test_buckets(self):
        records = [
            _make_video("v1", "A", "音乐", "2024-06-20 02:00:00", 600, 600),
            _make_video("v2", "B", "音乐", "2024-06-20 08:00:00", 600, 600),
            _make_video("v3", "C", "音乐", "2024-06-20 14:00:00", 600, 600),
            _make_video("v4", "D", "音乐", "2024-06-20 20:00:00", 600, 600),
        ]
        dist = time_of_day_distribution(records)
        assert dist == {"凌晨": 1, "上午": 1, "下午": 1, "晚上": 1}


class TestCompletion:
    def test_min_count_filter(self):
        records = [
            _make_video("v1", "A", "音乐", "2024-06-20 10:00:00", 600, 600),
            _make_video("v2", "A", "音乐", "2024-06-20 11:00:00", 600, 600),
            _make_video("v3", "B", "音乐", "2024-06-20 12:00:00", 600, 600),
        ]
        result = top_authors_by_completion(records, min_count=2)
        assert len(result) == 1
        assert result[0]["name"] == "A"


class TestDailyTrend:
    def test_recent_days(self):
        from datetime import datetime, timedelta
        today = datetime.now()
        records = [
            _make_video("v1", "A", "音乐", today.strftime("%Y-%m-%d %H:%M:%S"), 600, 600),
        ]
        trend = daily_trend(records, days=7)
        assert len(trend) == 7
        assert trend[-1]["count"] == 1
        assert trend[0]["count"] == 0

File: tests/business/test_analytics.cpp
#include <gtest/gtest.h>

#include "business/analytics.h"
#include "core/models.h"

#include <QDateTime>

using namespace bili;
using namespace bili::business;

namespace {

RecordPtr makeVideo(const QString& title,
                    const QString& author,
                    const QString& category,
                    int duration,
                    int progressPercent,
                    const QDateTime& viewAt)
{
    auto record = std::make_shared<VideoRecord>();
    record->type = RecordType::Video;
    record->title = title;
    record->authorName = author;
    record->category = category;
    record->duration = duration;
    record->progressPercent = progressPercent;
    record->viewAt = viewAt;
    return record;
}

RecordPtr makeLive(const QString& title,
                   const QString& author,
                   const QDateTime& viewAt)
{
    auto record = std::make_shared<LiveRecord>();
    record->type = RecordType::Live;
    record->title = title;
    record->authorName = author;
    record->category = QStringLiteral("直播");
    record->viewAt = viewAt;
    return record;
}

} // namespace

TEST(AnalyticsTest, ComputeBasicStats)
{
    RecordList records;
    const QDateTime now = QDateTime::currentDateTime();
    records.push_back(makeVideo(QStringLiteral("视频1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 50, now));
    records.push_back(makeVideo(QStringLiteral("视频2"), QStringLiteral("UP_A"), QStringLiteral("动画"), 200, 100, now));
    records.push_back(makeLive(QStringLiteral("直播1"), QStringLiteral("UP_B"), now));

    const QVariantMap stats = computeBasicStats(records);

    EXPECT_EQ(stats[QStringLiteral("total_records")].toInt(), 3);
    EXPECT_EQ(stats[QStringLiteral("total_videos")].toInt(), 2);
    EXPECT_EQ(stats[QStringLiteral("total_lives")].toInt(), 1);
    EXPECT_EQ(stats[QStringLiteral("total_articles")].toInt(), 0);
    EXPECT_EQ(stats[QStringLiteral("unique_authors")].toInt(), 2);
    EXPECT_EQ(stats[QStringLiteral("total_watch_seconds")].toLongLong(), 250);
    EXPECT_EQ(stats[QStringLiteral("avg_completion")].toInt(), 50);
}

TEST(AnalyticsTest, TopAuthors)
{
    RecordList records;
    const QDateTime now = QDateTime::currentDateTime();
    records.push_back(makeVideo(QStringLiteral("v1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v2"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v3"), QStringLiteral("UP_B"), QStringLiteral("动画"), 100, 50, now));

    const QVariantList authors = topAuthors(records, 10);
    ASSERT_EQ(authors.size(), 2);

    EXPECT_EQ(authors[0].toMap()[QStringLiteral("name")].toString(), QStringLiteral("UP_A"));
    EXPECT_EQ(authors[0].toMap()[QStringLiteral("count")].toInt(), 2);
    EXPECT_EQ(authors[1].toMap()[QStringLiteral("name")].toString(), QStringLiteral("UP_B"));
}

TEST(AnalyticsTest, TopCategories)
{
    RecordList records;
    const QDateTime now = QDateTime::currentDateTime();
    records.push_back(makeVideo(QStringLiteral("v1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v2"), QStringLiteral("UP_B"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v3"), QStringLiteral("UP_C"), QStringLiteral("音乐"), 100, 100, now));

    const QVariantList categories = topCategories(records, 10);
    ASSERT_EQ(categories.size(), 2);

    EXPECT_EQ(categories[0].toMap()[QStringLiteral("name")].toString(), QStringLiteral("动画"));
    EXPECT_EQ(categories[0].toMap()[QStringLiteral("count")].toInt(), 2);
}

TEST(AnalyticsTest, TimeOfDayDistribution)
{
    RecordList records;
    QDateTime morning = QDateTime::currentDateTime();
    morning.setTime(QTime(9, 0));
    QDateTime evening = QDateTime::currentDateTime();
    evening.setTime(QTime(21, 0));

    records.push_back(makeVideo(QStringLiteral("v1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, morning));
    records.push_back(makeVideo(QStringLiteral("v2"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, morning));
    records.push_back(makeVideo(QStringLiteral("v3"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, evening));

    const QVariantMap dist = timeOfDayDistribution(records);
    EXPECT_EQ(dist[QStringLiteral("上午")].toInt(), 2);
    EXPECT_EQ(dist[QStringLiteral("晚上")].toInt(), 1);
    EXPECT_EQ(dist[QStringLiteral("下午")].toInt(), 0);
}

TEST(AnalyticsTest, TopAuthorsByCompletion)
{
    RecordList records;
    const QDateTime now = QDateTime::currentDateTime();
    records.push_back(makeVideo(QStringLiteral("v1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v2"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v3"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, now));
    records.push_back(makeVideo(QStringLiteral("v4"), QStringLiteral("UP_B"), QStringLiteral("动画"), 100, 50, now));

    const QVariantList authors = topAuthorsByCompletion(records, 3, 10);
    ASSERT_EQ(authors.size(), 1);
    EXPECT_EQ(authors[0].toMap()[QStringLiteral("name")].toString(), QStringLiteral("UP_A"));
    EXPECT_EQ(authors[0].toMap()[QStringLiteral("avg_completion")].toInt(), 100);
}

TEST(AnalyticsTest, DailyTrend)
{
    RecordList records;
    QDateTime today = QDateTime::currentDateTime();
    QDateTime yesterday = today.addDays(-1);

    records.push_back(makeVideo(QStringLiteral("v1"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, today));
    records.push_back(makeVideo(QStringLiteral("v2"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, today));
    records.push_back(makeVideo(QStringLiteral("v3"), QStringLiteral("UP_A"), QStringLiteral("动画"), 100, 100, yesterday));

    const QVariantList trend = dailyTrend(records, 7);
    ASSERT_EQ(trend.size(), 7);

    const QVariantMap last = trend.last().toMap();
    EXPECT_EQ(last[QStringLiteral("count")].toInt(), 2);
}

TEST(AnalyticsTest, FormatDuration)
{
    EXPECT_EQ(formatDuration(0), QStringLiteral("0秒"));
    EXPECT_EQ(formatDuration(45), QStringLiteral("45秒"));
    EXPECT_EQ(formatDuration(125), QStringLiteral("2分5秒"));
    EXPECT_EQ(formatDuration(3665), QStringLiteral("1小时1分5秒"));
    EXPECT_TRUE(formatDuration(90061).contains(QStringLiteral("天")));
}
